## Question 1

Set, Dictionary

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib

eaf_data = pd.read_csv('FoodBalanceSheets_E_Africa_NOFLAG.csv', encoding='latin-1')
eaf_data

Using matplotlib backend: <object object at 0x116c9c970>


Area Code      Area  Item Code           Item  Element Code   
0              4   Algeria       2501     Population           511  \
1              4   Algeria       2501     Population          5301   
2              4   Algeria       2901    Grand Total           664   
3              4   Algeria       2901    Grand Total           674   
4              4   Algeria       2901    Grand Total           684   
...          ...       ...        ...            ...           ...   
60938        181  Zimbabwe       2899  Miscellaneous          5142   
60939        181  Zimbabwe       2899  Miscellaneous           645   
60940        181  Zimbabwe       2899  Miscellaneous           664   
60941        181  Zimbabwe       2899  Miscellaneous           674   
60942        181  Zimbabwe       2899  Miscellaneous           684   

                                      Element             Unit     Y2014   
0               Total Population - Both sexes     1000 persons  38924.00  \
1                    Domestic supply quantity      1000 tonnes      0.00   
2               Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3      Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4          Fat supply quantity (g/capita/day)     g/capita/day     80.06   
...                                       ...              ...       ...   
60938                                    Food      1000 tonnes     42.00   
60939     Food supply quantity (kg/capita/yr)               kg      3.06   
60940           Food supply (kcal/capita/day)  kcal/capita/day      3.00   
60941  Protein supply quantity (g/capita/day)     g/capita/day      0.10   
60942      Fat supply quantity (g/capita/day)     g/capita/day      0.04   

          Y2015     Y2016     Y2017     Y2018  
0      39728.00  40551.00  41389.00  42228.00  
1          0.00      0.00      0.00      0.00  
2       3379.00   3372.00   3341.00   3322.00  
3         94.35     94.72     92.82     91.83  
4         79.36     77.40     80.19     77.28  
...         ...       ...       ...       ...  
60938     46.00     33.00     19.00     16.00  
60939      3.33      2.35      1.33      1.08  
60940      4.00      3.00      1.00      1.00  
60941      0.11      0.08      0.04      0.04  
60942      0.05      0.03      0.02      0.01  

[60943 rows x 12 columns]

In [7]:
eaf_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60943 entries, 0 to 60942
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Area Code     60943 non-null  int64  
 1   Area          60943 non-null  object 
 2   Item Code     60943 non-null  int64  
 3   Item          60943 non-null  object 
 4   Element Code  60943 non-null  int64  
 5   Element       60943 non-null  object 
 6   Unit          60943 non-null  object 
 7   Y2014         59354 non-null  float64
 8   Y2015         59395 non-null  float64
 9   Y2016         59408 non-null  float64
 10  Y2017         59437 non-null  float64
 11  Y2018         59507 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 5.6+ MB


## Question 2

In [3]:
array = np.array([[94,89,63], 
          [93,92,48], 
          [92,94,56]])
array[:2,1:]

array([[89, 63],
       [92, 48]])

## Question 3

Which of these Areas had the highest sum in 2017?

In [5]:
eaf_data.groupby('Area').sum()['Y2017'].sort_values(ascending=False)

Area
Nigeria                        1483268.23
Egypt                           866379.92
South Africa                    517590.54
Ethiopia                        448683.76
Morocco                         388495.36
Ghana                           337599.06
Algeria                         325644.27
United Republic of Tanzania     322616.85
Kenya                           264660.66
Sudan                           239931.92
Cameroon                        232030.43
Angola                          229159.57
Côte d'Ivoire                   224599.01
Uganda                          213950.38
Malawi                          181098.71
Mozambique                      161407.98
Mauritania                      156665.46
Mali                            149928.33
Madagascar                      131197.73
Niger                           126707.58
Benin                           124771.22
Tunisia                         124167.20
Zambia                          103223.77
Burkina Faso                 

Nigeria had the highest sum in 2017

## Question 4

Which year had the least correlation with ‘Element Code’?

In [6]:
eaf_data.select_dtypes(include='number').corr()['Element Code'].sort_values(ascending=False)

Element Code    1.000000
Y2014           0.024457
Y2018           0.024279
Y2017           0.024254
Y2015           0.023889
Y2016           0.023444
Area Code      -0.000209
Item Code      -0.024683
Name: Element Code, dtype: float64

Y2016 had the least correlation with Element Code

## Question 5

c and d

## Question 6

What is the total number and percentage of missing data in 2014 to 3 decimal places?

In [18]:
Total = round(eaf_data['Y2014'].isna().sum(), 3)
percentage = round(Total*100/len(eaf_data), 3)
print(f"Total number and percentage of missing data in 2014 is: {Total} and {percentage} respectively.")

Total number and percentage of missing data in 2014 is: 1589 and 2.607 respectively.


## Question 7

In [9]:
eaf_data.shape

(60943, 12)

## Question 8

100 features, 3 observations.

## Question 9

Which of these Areas had the 7th lowest sum in 2017?

In [10]:
eaf_data.groupby('Area').sum()['Y2017'].sort_values(ascending=False)

Area
Nigeria                        1483268.23
Egypt                           866379.92
South Africa                    517590.54
Ethiopia                        448683.76
Morocco                         388495.36
Ghana                           337599.06
Algeria                         325644.27
United Republic of Tanzania     322616.85
Kenya                           264660.66
Sudan                           239931.92
Cameroon                        232030.43
Angola                          229159.57
Côte d'Ivoire                   224599.01
Uganda                          213950.38
Malawi                          181098.71
Mozambique                      161407.98
Mauritania                      156665.46
Mali                            149928.33
Madagascar                      131197.73
Niger                           126707.58
Benin                           124771.22
Tunisia                         124167.20
Zambia                          103223.77
Burkina Faso                 

Guinea-Bissau had the 7th lowest sum in 2017 

## Question 10

In [11]:
my_tuppy = (1,2,5,8)
my_tuppy[2] = 6

TypeError: 'tuple' object does not support item assignment

Type Error

## Question 11

In [12]:
y = [(2,4), (7,8), (1,5,9)]
x = y[1][-1]
print(x)

8


In [13]:
y = [(2,4), (7,8), (1,5,9)]
x = y[1][1]
print(x)

8


x = y[1][-1] and x = y[1][1]

## Question 12

What is the total sum of Wine produced in 2015 and 2018 respectively?

In [14]:
eaf_data.groupby(['Item']).sum().loc['Wine']

Area Code                                                   66206
Area            AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...
Item Code                                                 1319535
Element Code                                              1818328
Element         ProductionImport QuantityStock VariationExport...
Unit            1000 tonnes1000 tonnes1000 tonnes1000 tonnes10...
Y2014                                                     4497.36
Y2015                                                     4251.81
Y2016                                                     3872.09
Y2017                                                     4178.02
Y2018                                                     4039.32
Name: Wine, dtype: object

The total sum of Wine produced in 2015 and 2018 is: 4251.81 and 4039.32 respectively

## Question 13

Math

## Question 14

What is the total Protein supply quantity in Madagascar in 2015?

In [22]:
eaf_data.query("Area == 'Madagascar'").groupby('Element')['Y2015'].sum()

Element
Domestic supply quantity                  31214.98
Export Quantity                             494.75
Fat supply quantity (g/capita/day)           91.85
Feed                                       2070.22
Food                                      21120.65
Food supply (kcal/capita/day)              7685.00
Food supply quantity (kg/capita/yr)         871.59
Import Quantity                            1721.80
Losses                                     2176.00
Other uses (non-food)                      3957.12
Processing                                 1803.00
Production                                29482.89
Protein supply quantity (g/capita/day)      173.05
Residuals                                  -616.00
Seed                                        699.00
Stock Variation                            -505.00
Total Population - Both sexes             24234.00
Name: Y2015, dtype: float64

The total Protein supply quantity in Madagascar in 2015 is: 173.05 

## Question 15

In [15]:
lst = [[35, 'Portugal',94],[33,'Argentina',93],[30,'Brazil',92]]
col = ['Age','Nationality','Overall']
p_df = pd.DataFrame(lst, columns = col, index=[1,2,3])
p_df

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

Answer is: None of the above

## Question 16

In [16]:
S = [['him', 'sell'], [90, 28, 43]]
S[0][1][1]

'e'

## Question 17

What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [17]:
mean = round(eaf_data['Y2017'].mean(), 2)
std = round(eaf_data['Y2017'].std(), 2)
print(f"The mean and standard deviation across the whole dataset for the year 2017 is: {mean} and {std}")

The mean and standard deviation across the whole dataset for the year 2017 is: 140.92 and 1671.86


## Question 18

What year has the highest sum of Stock Variation?

In [19]:
eaf_data.groupby('Element').sum().loc['Stock Variation']['Y2014':].sort_values(ascending=False)

Y2014    58749.83
Y2017    54316.91
Y2015    34910.99
Y2016    33140.12
Y2018    20577.91
Name: Stock Variation, dtype: object

Y2014 has the highest sum of Stock Variation

## Question 19

What is the total number of unique countries in the dataset?

In [20]:
eaf_data['Area'].nunique()

49

## Question 20

What is the total number of the sum of Processing in 2017?

In [21]:
eaf_data.groupby('Element')[['Y2017']].sum().loc['Processing']

Y2017    292836.0
Name: Processing, dtype: float64